File: sdks/typescript/src/resources/webhooks.ts
import { AxiosInstance } from 'axios';
import { PaginatedResponse } from '../types';

export interface Webhook {
    id: string;
    url: string;
    events: string[];
    is_active: boolean;
    secret?: string;
    created_at: string;
    updated_at: string;
}

export interface WebhookEvent {
    id: string;
    webhook: string;
    event_type: string;
    payload: any;
    status_code?: number;
    response?: string;
    created_at: string;
}

export class Webhooks {
    constructor(private client: AxiosInstance) { }

    async list(page: number = 1, pageSize: number = 20): Promise<PaginatedResponse<Webhook>> {
        const response = await this.client.get('/webhooks/', {
            params: { page, page_size: pageSize }
        });
        return response.data;
    }

    async get(webhookId: string): Promise<Webhook> {
        const response = await this.client.get(`/webhooks/${webhookId}/`);
        return response.data;
    }

    async create(url: string, events: string[], secret?: string): Promise<Webhook> {
        const data: any = { url, events };
        if (secret) data.secret = secret;
        const response = await this.client.post('/webhooks/', data);
        return response.data;
    }

    async update(webhookId: string, data: Partial<Webhook>): Promise<Webhook> {
        const response = await this.client.patch(`/webhooks/${webhookId}/`, data);
        return response.data;
    }

    async delete(webhookId: string): Promise<void> {
        await this.client.delete(`/webhooks/${webhookId}/`);
    }

    async listEvents(webhookId: string, page: number = 1): Promise<PaginatedResponse<WebhookEvent>> {
        const response = await this.client.get('/webhook-logs/', {
            params: { webhook: webhookId, page }
        });
        return response.data;
    }
}

File: sdks/rust/src/resources/webhooks.rs
use crate::models::{PaginatedResponse, Webhook};
use crate::Puxbay;
use crate::PuxbayError;
use reqwest::Method;
use std::collections::HashMap;

pub struct WebhooksResource<'a> {
    pub(crate) client: &'a Puxbay,
}

impl<'a> WebhooksResource<'a> {
    pub fn new(client: &'a Puxbay) -> Self {
        Self { client }
    }

    pub async fn list(&self, page: i32) -> Result<PaginatedResponse<Webhook>, PuxbayError> {
        self.client.request(
            Method::GET,
            &format!("webhooks/?page={}", page),
            None::<()>,
        ).await
    }

    pub async fn get(&self, webhook_id: &str) -> Result<Webhook, PuxbayError> {
        self.client.request(
            Method::GET,
            &format!("webhooks/{}/", webhook_id),
            None::<()>,
        ).await
    }

    pub async fn create(&self, webhook: &Webhook) -> Result<Webhook, PuxbayError> {
        self.client.request(
            Method::POST,
            "webhooks/",
            Some(webhook),
        ).await
    }

    pub async fn update(&self, webhook_id: &str, webhook: &Webhook) -> Result<Webhook, PuxbayError> {
        self.client.request(
            Method::PATCH,
            &format!("webhooks/{}/", webhook_id),
            Some(webhook),
        ).await
    }

    pub async fn delete(&self, webhook_id: &str) -> Result<(), PuxbayError> {
        self.client.request::<serde_json::Value>(
            Method::DELETE,
            &format!("webhooks/{}/", webhook_id),
            None::<()>,
        ).await?;
        Ok(())
    }

    pub async fn list_events(&self, webhook_id: &str, page: i32) -> Result<HashMap<String, serde_json::Value>, PuxbayError> {
        self.client.request(
            Method::GET,
            &format!("webhook-logs/?webhook={}&page={}", webhook_id, page),
            None::<()>,
        ).await
    }
}
